File: lib/availability.ts
export const openingHours = {
  weekdayNightStartHour: 18,
  weekdayNightEndHour: 4,
  slotStepMin: 30,
  timeZone: "Europe/Bucharest"
};

export function getEndDate(startsAt: Date, durationMin: number) {
  return new Date(startsAt.getTime() + durationMin * 60 * 1000);
}

export function hasOverlap(
  startsAt: Date,
  endsAt: Date,
  item: { startsAt: Date; endsAt: Date }
) {
  return startsAt < item.endsAt && endsAt > item.startsAt;
}

export function isInsideBusinessHours(startsAt: Date, endsAt: Date) {
  if (endsAt <= startsAt) {
    return false;
  }

  const sampleStepMs = openingHours.slotStepMin * 60 * 1000;

  for (
    let timestamp = startsAt.getTime();
    timestamp < endsAt.getTime();
    timestamp += sampleStepMs
  ) {
    if (!isOpenAt(new Date(timestamp))) {
      return false;
    }
  }

  return isOpenAt(new Date(endsAt.getTime() - 60 * 1000));
}

export function isOpenAt(date: Date) {
  const local = getZonedParts(date);
  const hour = local.hour + local.minute / 60;
  const isWeekend = local.weekday === "Sat" || local.weekday === "Sun";

  if (isWeekend) {
    return true;
  }

  return hour >= openingHours.weekdayNightStartHour || hour < openingHours.weekdayNightEndHour;
}

function getZonedParts(date: Date) {
  const parts = new Intl.DateTimeFormat("en-US", {
    timeZone: openingHours.timeZone,
    weekday: "short",
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
    hour: "2-digit",
    minute: "2-digit",
    hour12: false
  }).formatToParts(date);

  const value = (type: string) => parts.find((part) => part.type === type)?.value ?? "";

  return {
    weekday: value("weekday"),
    year: Number(value("year")),
    month: Number(value("month")),
    day: Number(value("day")),
    hour: Number(value("hour")),
    minute: Number(value("minute"))
  };
}

File: backend/app/availability.py
from __future__ import annotations

from datetime import datetime, timedelta
from zoneinfo import ZoneInfo

TIMEZONE = ZoneInfo("Europe/Bucharest")
WEEKDAY_NIGHT_START = 18
SAMPLE_STEP_MINUTES = 30


def get_end_date(starts_at: datetime, duration_min: int) -> datetime:
    return starts_at + timedelta(minutes=duration_min)


def is_open_at(moment: datetime) -> bool:
    local = moment.astimezone(TIMEZONE)
    return local.hour >= WEEKDAY_NIGHT_START


def is_inside_business_hours(starts_at: datetime, ends_at: datetime) -> bool:
    if ends_at <= starts_at:
        return False

    current = starts_at
    step = timedelta(minutes=SAMPLE_STEP_MINUTES)
    while current < ends_at:
        if not is_open_at(current):
            return False
        current += step

    return is_open_at(ends_at - timedelta(seconds=60))
